internal/schemautil/schemautil.go

```diff
@@ -382,88 +386,40 @@ func ResourceDataSet(d *schema.ResourceData, s map[string]*schema.Schema, dto an
 		return err
 	}
 
-	result, err := serializeMap(s, m)
-	if err != nil {
-		return err
+	for _, f := range fns {
+		for k, v := range m {
+			delete(m, k) // remove the old key in case it is replaced
+			k, v = f(k, v)
+			m[k] = v
+		}
 	}
 
-	for k, v := range result {
-		if err = d.Set(k, v); err != nil {
-			return err
+	for k := range s {
+		if v, ok := m[k]; ok {
+			if err = d.Set(k, v); err != nil {
+				return err
+			}
 		}
 	}
 	return nil
 }
 
-func serializeMap(s map[string]*schema.Schema, m map[string]any) (map[string]any, error) {
-	result := make(map[string]any)
-	for name, field := range s {
-		value, ok := m[name]
-		if !ok {
-			continue
-		}
-
-		val, err := serialize(field, value)
-		if err != nil {
-			return nil, err
+// RenameAliases renames field names on object top level
+func RenameAliases(aliases map[string]string) KVModifier {
+	return func(k string, v any) (string, any) {
+		alias, ok := aliases[k]
+		if ok {
+			return alias, v
 		}
-		result[name] = val
+		return k, v
 	}
-
-	return result, nil
 }
 
-func serialize(s *schema.Schema, value any) (any, error) {
-	switch s.Type {
-	case schema.TypeList, schema.TypeSet:
-	default:
-		return value, nil
-	}
-
-	var err error
-	list, isList := value.([]any)
-
-	switch elem := s.Elem.(type) {
-	case *schema.Schema:
-		// This branch converts a list of scalars
-		if !isList {
-			return nil, fmt.Errorf("expected a list, but %T", value)
-		}
-
-		result := make([]any, len(list))
-		for i, v := range list {
-			result[i], err = serialize(elem, v)
-			if err != nil {
-				return nil, err
-			}
-		}
-
-		return schema.NewSet(schema.HashSchema(elem), result), nil
-	case *schema.Resource:
-		// This branch converts a list of objects or a single object
-		if m, ok := value.(map[string]any); ok {
-			return serializeMap(elem.Schema, m)
-		}
-
-		if !isList {
-			return nil, fmt.Errorf("expected a map or a list, but %T", value)
-		}
-
-		result := make([]any, len(list))
-		for i, v := range list {
-			m, ok := v.(map[string]any)
-			if !ok {
-				return nil, fmt.Errorf("expected a map, got %T", v)
-			}
-
-			result[i], err = serializeMap(elem.Schema, m)
-			if err != nil {
-				return nil, err
-			}
-		}
-		return result, nil
```

Comment:
It turns out that the TF behaviour on top level fields is different compared to the nested fields. This code was copied from user config convertors. It might happen to be a case later, but for these resources TF does not accept TypeSet values.